### ISSUE TITLE
Use dlmopen for modules to clearly separate symbol namespace

### ISSUE THREAD

**Issue by jaromil** [tarantool/tarantool#4746]
Is this the `dlopen` call for loaded modules? https://github.com/tarantool/tarantool/blob/ca07088cd4ad45f04951504e0a5924af2de0129c/src/box/func.c#L276

I'm wondering if you are open to a change to the way you load modules, to adopt `dlmopen` as I've described here https://github.com/antirez/redis/pull/6125

This change would allow modules to carry their own symbols without clashing with tarantool's.

For instance modules with their own modified Lua interpreter.

**Comment by cyrillos** [comment]
Actually it looks tempting to support but we need more time to figure out the details and potential side effects.

**Comment by cyrillos** [comment]
Thinking more I fear this might break backward compatibility. What if there already present users/modules which relies on global symbol namespace?

**Comment by jaromil** [comment]
Yes, glibc's `dlmopen` provides complete isolation; but fallback can be handled by the host application.

I see a solution here and I wish also @antirez would consider it for Redis, I'm speaking based on experience made maintaining frei0r and following LADSPA, both large third-party communities of plugin developers.

The solution can be to add metadata that indicates loading options (and more) for plugins.
Such metadata can be either an external file (LADSPA uses rdf) that goes side-by-side with the binary DSO or a function symbol of the DSO to be called after an initial `dlopen` and returning a struct of options.

Then the flow can be:
1. read metadata (file load, or dlopen+setup() call)
2. if setup() returns full isolation option, dlclose() and dlmopen() the DSO
3. pass the DSO pointer and process

There may be more solutions, in any case I recommend you do consider the isolation flag as a desirable option for plugin developers.

**Comment by cyrillos** [comment]
Yes, I think of some kind of flag to provide an option for such load, but definitely we won't switch to `dlmopen` by default.

**Comment by jaromil** [comment]
Ping. Any thoughts about this? A runtime configuration switch?

**Comment by cyrillos** [comment]
Hi! Sorry but have had no time to think about, -EBUSY :(
